File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

public class Solution {
    public static int[][] dp = new int[501][501];

    public static int solution(int[][] triangle) {
        for(int[] d : dp) {
            Arrays.fill(d, -1);
        }

        return max(0, 0, triangle);
    }

    public static int max(int x, int y, int[][] triangle) {
        if(y == triangle.length) {
            return 0;
        }

        if(dp[y][x] != -1) {
            return dp[y][x];
        }

        return dp[y][x] = triangle[y][x] + Math.max(max(x, y + 1, triangle), max(x + 1, y + 1, triangle));
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <cmath>
using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    vector<vector<int>> dp = triangle;
    
    for(int i = 1; i < dp.size(); ++i) {
        for(int j = 0; j <= i; ++j) {
            int left = j - 1;
            int right = j;
            
            if (left < 0) {
                dp[i][j] += dp[i-1][right];
            } else if (right > i-1) {
                dp[i][j] += dp[i-1][left];
            } else {
                dp[i][j] += max(dp[i-1][right], dp[i-1][left]);
            }
        }
    }
    for(int i = 0; i < dp.back().size(); ++i) {
        answer = max(answer, dp.back()[i]);
    }
    return answer;
}